mipengine/controller/algorithm_executor.py

```diff
@@ -294,12 +249,75 @@ def metadata(self):
 
     @property
     def datasets_per_local_node(self):
-        return self.datasets_per_local_node
+        return self._datasets_per_local_node
 
     @property
     def use_smpc(self):
         return self._use_smpc
 
+    def create_primary_data_views(
+        self,
+        variable_groups: List[List[str]],
+        dropna: bool = True,
+        check_min_rows: bool = True,
+    ) -> List[LocalNodesTable]:
+        """
+        Creates primary data views, for each variable group provided,
+        using also the algorithm request arguments (data_model, datasets, filters).
+
+        Parameters
+        ----------
+        variable_groups : List[List[str]]
+            A list of variable_groups. The variable group is a list of columns.
+        dropna : bool
+            Remove NAs from the view.
+        check_min_rows : bool
+            Raise an exception if there are not enough rows in the view.
+
+        Returns
+        ------
+        List[LocalNodesTable]
+            A (LocalNodesTable) view for each variable_group provided.
+        """
+
+        command_id = str(get_next_command_id())
+        views_per_localnode = [
+            (
+                local_node,
+                local_node.create_data_model_views(
+                    command_id=command_id,
+                    data_model=self._data_model,
+                    datasets=self.datasets_per_local_node[local_node.node_id],
+                    columns_per_view=variable_groups,
+                    filters=self._var_filters,
+                    dropna=dropna,
+                    check_min_rows=check_min_rows,
+                ),
+            )
+            for local_node in self._local_nodes
+        ]
+
+        # Add empty dicts for as many views created in the nodes
+        views_of_first_node = views_per_localnode[0][1]
+        local_nodes_tables_dicts = [{} for _ in views_of_first_node]
+        for localnode, localnode_views in views_per_localnode:
+            if len(localnode_views) != len(local_nodes_tables_dicts):
+                assert ValueError(
+                    f"All views from localnodes should have the same length. "
+                    f"{localnode_views} and {local_nodes_tables_dicts} have different length."
+                )
+
+            # Each view in a node should be added to the proper local_nodes_tables_dict
+            for view, local_nodes_tables in zip(
+                localnode_views, local_nodes_tables_dicts
+            ):
+                local_nodes_tables[localnode] = view
+
+        return [
+            LocalNodesTable(local_nodes_tables_dict)
+            for local_nodes_tables_dict in local_nodes_tables_dicts
+        ]
```

Comment:
As a general comment, this method contains the main change in the logic, it is the main reason for this pr, it should not be so convoluted, it should have been written in a much simpler way.

Reply:
this is even worse than the previous one